Add Turn Touch component.

antsar opened the pull request:
## Description

Implements a component for the [Turn Touch](https://shop.turntouch.com/) bluetooth remote control using the [TurnTouch library](https://pypi.org/project/TurnTouch/).

This provides one component (`turn_touch`) which fires button press events, and one platform (`sensor.turn_touch`) which exposes the remote's battery level. The platform is configured automatically by the component (following [Tesla](https://www.home-assistant.io/components/tesla/) as an example).

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation:** home-assistant/home-assistant.github.io#5591

## Example entry for `configuration.yaml` (if applicable):
```yaml
turn_touch:
  devices:
    - mac: '00:de:ad:be:ef:00'
      debounce: false   # bypass 0.75 second delay at the expense of double-tap & hold
    - mac: 'c0:ff:ee:c0:ff:ee'
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] ~New files were added to `.coveragerc`.~ No need; tests included.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

antsar committed:
Add Turn Touch component.

pvizeli commented on homeassistant/components/sensor/turn_touch.py:
```diff
@@ -0,0 +1,67 @@
+        """
+        _LOGGER.debug('Initializing turn touch battery sensor.')
+        self.turn_touch = turntouch_device
+        self.entity_id = ENTITY_ID_FORMAT.format(self.turn_touch.unique_id)
```
Remove this and make a unique_id property with that value

antsar commented:
Thanks for the quick feedback @pvizeli! Code has been updated.

pvizeli commented on homeassistant/components/turn_touch.py:
```diff
@@ -0,0 +1,178 @@
+    @property
+    def unique_id(self):
+        """Return the Unique ID (based on MAC address) of this remote."""
+        return 'TurnTouch_{mac}'.format(mac=self._address.replace(':', ''))
```
only return the mac

pvizeli commented on homeassistant/components/sensor/turn_touch.py:
```diff
@@ -0,0 +1,71 @@
+import logging
+
+from homeassistant.const import DEVICE_CLASS_BATTERY
+from homeassistant.components import turn_touch
```
Only import data key from turn_touch

pvizeli commented on homeassistant/components/sensor/turn_touch.py:
```diff
@@ -0,0 +1,71 @@
+        """
+        _LOGGER.debug('Initializing turn touch battery sensor.')
+        self.turn_touch = turntouch_device
+        self._battery_level = self.turn_touch.get_battery()
```
move this into `def update(self)` and call add_devices(..., True)

antsar replied:
@pvizeli Just one question about your comment:

> Remove this reconnect logic from functions

Those functions aren't called from `@property` methods (`name()`, `state()`), only from `TurnTouchRemote__init_.()` and `TurnTouchBatterySensor.update()`. My understanding is that Home Assistant initializes components and calls `update()` from a separate thread already, so wrapping it in another thread seems redundant. Is that wrong?

pvizeli reviewed (changes requested):
Remove this reconnect logic from functions. You can start a thread and manage the device connections, but if home assistant ask for data, return the available data and don't try to realod it.

MartinHjelmare commented on homeassistant/components/turn_touch.py:
```diff
@@ -0,0 +1,178 @@
+    hass.data[DATA_KEY] = {'devices': {}}
+
+    for device_config in config[DOMAIN][CONF_DEVICES]:
+        turn_touch = TurnTouchRemote(hass, device_config)
```
Entity instances should not be created by the component but by the platform. Move this to the platform.

antsar replied:
Thanks for the feedback @MartinHjelmare. I'm a bit confused about how this would work, and I'd appreciate some more guidance.

I initialize the device connection from the component because this is a bluetooth device and we can only have a single connection to it. If the device were (hypothetically) extended to support more than one type of sensor (lets say battery sensor + humidity sensor), how would that work if the entity is defined/initialized in each of the platforms?

Right now there is only one platform (the battery sensor), but even so, the real/primary device is the remote, not the battery sensor. So it seems logical to represent (& connect to) the remote in the component, and use the platform only for the battery level part.

MartinHjelmare replied:
Separate the entity from the connection and keep the connection in the component and the entity creation in each platform.

MartinHjelmare reviewed (changes requested):
Separate the entity from the event firing. Event firing should be done by the component but the entity should be created and contained in the platform.

antsar commented:
Thanks @pvizeli and @MartinHjelmare for your feedback. I've made the suggested changes with one exception:

> 
> @pvizeli Just one question about your comment:
> 
>     Remove this reconnect logic from functions
> 
> Those methods (`_connect()`, `_listen()`, `get_name()`, `get_battery()`) aren't called from `@property` methods (`name()`, `state()`), only from `TurnTouchRemote__init_.()` and `TurnTouchBatterySensor.update()`. My understanding is that Home Assistant initializes components and calls `update()` from a separate thread already, so wrapping it in another thread seems redundant. Is that wrong?
> 
> If so, any chance there's any example code you can suggest for the best way to handle this?

Here are the changes I've made:

- Don't inherit from `Entity` in the component; only in the platform.
- Don't use `ENTITY_ID_FORMAT` in `unique_id`.
- Don't get initial sensor value in `__init__`; call `add_devices(..., True)` instead.
- Remove extra imports from the component into the sensor.
- Return only MAC address as `unique_id`, rather than prefixing with `TurnTouch_`.

MartinHjelmare commented on homeassistant/components/turn_touch.py:
```diff
@@ -0,0 +1,166 @@
+                return self.name
+            except turntouch.TurnTouchException:
+                _LOGGER.warning('Reading device name failed. Retrying...')
+            time.sleep(delay)
```
Sleeping in functions executed in the home assistant tread pool is not allowed. `get_name` is called inside component setup, ie the thread pool. It's ok to sleep in a separate thread. Async sleep is also ok.

MartinHjelmare commented on homeassistant/components/turn_touch.py:
```diff
@@ -0,0 +1,166 @@
+                return self._battery
+            except turntouch.TurnTouchException:
+                _LOGGER.warning('Reading device battery failed. Retrying...')
+            time.sleep(delay)
```
See above.

MartinHjelmare commented on homeassistant/components/turn_touch.py:
```diff
@@ -0,0 +1,166 @@
+                return True
+            except turntouch.TurnTouchException:
+                _LOGGER.warning('Connecting failed. Retrying...')
+            time.sleep(delay)
```
See above.

MartinHjelmare commented:
Please don't squash commits after review has started.

antsar committed:
Remove blocking IO from component setup.

Moved the _connect, get_name, and _listen calls into a new _setup method, which runs in a separate thread. Added locks to prevent concurrent IO from different threads (bluepy does not support that).

antsar committed:
Fix typo; improve coverage.

MartinHjelmare commented on homeassistant/components/turn_touch.py:
```diff
@@ -0,0 +1,189 @@
+            except AttributeError:
+                _LOGGER.warning('Cannot read battery level before connecting'
+                                ' to the device. Retrying...')
+            time.sleep(delay)
```
`get_battery` is still called from the entity `update` method, so sleeping here is not ok.

antsar replied:
Thanks for your patience and helpful feedback. `update` now kicks off a new thread for the `get_battery` call. I've also moved the `io_lock` (thread safety) stuff into the `TurnTouch` library, since that isn't a Home Assistant problem.

antsar committed:
Move all IO to separate threads.

bluepy IO is not reliable and may need to be retried. We use exponential backoff there. To avoid blocking Home Assistant threads with those delays, all IO is now performed in separate threads.

antsar committed:
Fix CI.
pylint throws E1111 for Threading.Lock.acquire(). This is incorrect, per Python docs.

See https://docs.python.org/3.7/library/threading.html#threading.Lock.acquire

MartinHjelmare reviewed (commented):
How do we protect from multiple threads spawning from the same entity if the connection never succeeds?

antsar committed:
Prevent multiple get_battery/get_name threads from piling up if connection is failing.

MartinHjelmare reviewed (commented):
I think we can make the code simpler if we only interface with the device from the single thread that also listens for events. That tread should handle reconnecting and reading battery and name at an interval, saving battery and name into instance attributes. Then entities just need to access the battery and name values of the instance attributes. So no locking or extra threads should be required.

antsar commented:
Thanks again for your reviews and helpful feedback, @MartinHjelmare! I have some concerns about that approach. Bear with me...

* `_listen` is blocking (this is a `bluepy` limitation), so we would need at least two threads: one for listening, one for getting name/battery. 
* `get_name` is only called once, when the device is initialized. The device can only be connected to one host at a time. As long as it's connected to Home Assistant, nothing should cause its name to change. So there's no need to call `get_name` after setup.
* `get_battery` *is* called repeatedly, but it is typically quick, and happens intermittently. In fact, I should add a `@Throttle(3600)` because battery level really does not change often.

So, I think the options are:

1. (currently implemented) Persistent thread for `_listen`, and an intermittent thread for `get_battery`.
2. Persistent thread for `_listen`, and another persistent thread for `get_battery`.

I could remove the locking code from `get_name` since it is only called once, but I thought it made sense to use a similar design for `get_name` and `get_battery` in case we do other stuff with `get_name` later (like allow the user to change it).

Does that make sense? I'm happy to make further changes, just want to make sure it makes sense for this device.

Thanks again for your help!

MartinHjelmare commented:
Ok. Then instead of a persistent thread for the updates we could schedule the spawning of the update threads, using our event helpers, from the component instead of from each entity. The update thread stores the latest state in instance attributes on the main data object. Then we wouldn't need the locks and we wouldn't have another persistent thread alive.
